homeassistant/helpers/entity_registry.py

```diff
@@ -616,6 +617,44 @@ def async_update_entity_options(
 
         return new
 
+    @callback
+    def async_migrate_entity_to_new_platform(
+        self,
+        entity_id: str,
+        new_platform: str,
+        *,
+        new_config_entry_id: str | UndefinedType = UNDEFINED,
+        new_unique_id: str | UndefinedType = UNDEFINED,
+        new_device_id: str | None | UndefinedType = UNDEFINED,
+    ) -> RegistryEntry:
+        """
+        Migrate entity to new platform.
+
+        This should only be used when an entity needs to be migrated between
+        integrations.
+        """
+        if not self.async_get(entity_id):
+            raise ValueError(f"Entity {entity_id} is not in the entity registry")
+        if (
+            state := self.hass.states.get(entity_id)
+        ) is not None and state.state != STATE_UNKNOWN:
+            raise ValueError("Only entities that haven't been loaded can be migrated")
+
+        old = self.entities[entity_id]
+        if new_config_entry_id == UNDEFINED and old.config_entry_id is not None:
+            raise ValueError(
+                f"new_config_entry_id required because {entity_id} is already linked "
+                "to a config entry"
+            )
+
+        new = self.entities[entity_id] = attr.evolve(old, platform=new_platform)
```

Comment:
There won't be an `EVENT_ENTITY_REGISTRY_UPDATED` fired for this change, not sure if that's acceptable.
If it's acceptable, we should add a comment here.
If it's not acceptable, we could either just teach `async_update_entity` about changing platforms, or
add an `_async_update_entity` which is called by this function and by `async_update_entity`.

Reply:
I created a new private function which is used by `async_update_entity`, `async_update_entity_options`, and `async_migrate_entity_to_new_platform` so that all logic is centralized